walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,93 @@
+    Collection<TimeRange> results = new ArrayList<>();
+    Collection<String> allAttendees = request.getAttendees();
+    int numAttendees = allAttendees.size();
+    long meetingTime = request.getDuration();
+
+    // Base Case: No required attendees.
+    if (numAttendees == 0) {
+        results = Arrays.asList(TimeRange.WHOLE_DAY);
+        return results;
+    // Base Case: Invalid meeting time.
+    } else if (meetingTime == 0 || meetingTime > TimeRange.WHOLE_DAY.duration()) {
+        results = Arrays.asList();
+        return results;
+    }
+
+    // Extract and sort event times (i.e. busy times) by start time.
+    List<TimeRange> timeRanges = new ArrayList<>();
+    for (Event e : events) {
+        // Ignore event times of people not actually attending.
+        boolean isOptional = isEventOptional(e, allAttendees);
+        if (!isOptional) {
+            timeRanges.add(e.getWhen());
+        }
+    }
+    timeRanges.sort(TimeRange.ORDER_BY_START);
+
+    // Track latest end time of any event.
+    // Since timeRanges was sorted by start, the event with the latest end time
+    // might come before the last event in the List (e.g. if we had [5, 20]
+    // followed by [10, 15], we want to use [20, END_OF_DAY] for the final gap,
+    // not [15, END_OF_DAY]).
+    int endOfLatestEvent = 0;
+    TimeRange gap;
+
+    // Iteratively construct gaps between busy times.
+    for (int i = 0; i < timeRanges.size(); i++) {        
+        // First gap before the first time anyone is busy.
+        if (i == 0) {
+            gap = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timeRanges.get(i).start(), false);
+        // gaps in the middle.
+        } else {
+            gap = TimeRange.fromStartEnd(timeRanges.get(i-1).end(), timeRanges.get(i).start(), false);
+        }
+
+        // Prevent edge behavior since start times are monotonically increasing,
+        // but end times are not.
+        if (isValidGap(gap, meetingTime)) {
+            results.add(gap);
+        }
+
+        // Update latest end time of any event.
+        if (i != timeRanges.size()) {
+            endOfLatestEvent = endOfLatestEvent > timeRanges.get(i).end() ? endOfLatestEvent : timeRanges.get(i).end();
+        }
+    }
+    // Last gap after the last person is busy.
+    gap = TimeRange.fromStartEnd(endOfLatestEvent, TimeRange.END_OF_DAY, true);
+    if (isValidGap(gap, meetingTime)) {
+        results.add(gap);
+    }
+
+    return results;
+
+  }
+
+  // Return `true` if a gap is "well-constructed" / valid:
+  // 1) Start comes before end
+  // 3) gap is big enough to have meeting
+  private boolean isValidGap(TimeRange gap, long meetingTime) {
+    return gap.start() < gap.end()
+        && gap.duration() >= meetingTime;
+    }
+
+  // Return `true` if event is optional i.e. none of the required
+  // attendees must go to the event. This means that gaps can be scheduled
+  // during the event.
```

Comment:
That's a really good point because one of the tests deals with considering/not considering optional events, and I think that's a good approach for me to try out.